pommerman/forward_model.py

```diff
@@ -156,128 +156,256 @@ def step(actions,
+            current_x, current_y = current
+            desired_x, desired_y = desired
+            if current_x != desired_x:
+                assert current_y == desired_y
+                return ('X', min(current_x, desired_x), current_y)
+            assert current_x == desired_x
+            return ('Y', current_x, min(current_y, desired_y))
+
+        for agent_num, agent in enumerate(alive_agents):
+            if desired_agent_positions[agent_num] != agent.position:
+                border = crossing(agent.position,
+                                  desired_agent_positions[agent_num])
+                if border in crossings:
+                    desired_agent_positions[agent_num] = agent.position
+                    num, isAgent = crossings[border]
+                    assert isAgent
+                    desired_agent_positions[num] = alive_agents[num].position
                 else:
-                    # The agent made an invalid direction.
-                    agent.stop()
-            else:
-                next_positions[agent.agent_id] = None
+                    crossings[border] = (agent_num, True)
+
+        for bomb_num, bomb in enumerate(curr_bombs):
+            if desired_bomb_positions[bomb_num] != bomb.position:
+                border = crossing(bomb.position,
+                                  desired_bomb_positions[bomb_num])
+                if border in crossings:
+                    desired_bomb_positions[bomb_num] = bomb.position
+                    num, isAgent = crossings[border]
+                    if not isAgent:
+                        desired_bomb_positions[num] = curr_bombs[num].position
+                else:
+                    crossings[border] = (bomb_num, False)
+
+        # Deal with multiple agents or multible bomb collisions on desired next position
+        # by resetting desired position to current position for everyone involved in the collision
+        agent_occupancy = np.zeros_like(curr_board)
+        bomb_occupancy = np.zeros_like(curr_board)
+
+        for desired_position in desired_agent_positions:
+            agent_occupancy[desired_position] += 1
+
+        for desired_position in desired_bomb_positions:
+            bomb_occupancy[position] += 1
+
+        # Resolve two or more agents or two or more bombs trying
+        # to occupy the same space.
+        # We will deal with an agent and a bomb trying to share space later
+        change = True
+        while change:
+            change = False
+            for agent_num, agent in enumerate(alive_agents):
+                desired_position = desired_agent_positions[agent_num]
+                curr_position = agent.position
+                if desired_position != curr_position and \
+                      (agent_occupancy[desired_position] > 1 or bomb_occupancy[desired_position] > 1):
+                    desired_agent_positions[agent_num] = curr_position
+                    agent_occupancy[curr_position] += 1
```

Comment:
shouldn't we also do:
```agent_occupancy[desired_position] -= 1```

Reply:
The idea is to keep collision locations marked.
Otherwise the location may be used in the kick stage.
( Two agents try to get onto the same square - bounce back and then a third agent kicks a bomb into the square (

Reply:
Ahh ok, that makes sense.

Reply:
I also considered a collision map - set to True if any bounce back happened on a square.
This could be exported for better visualization and we could keep exact counts ( also we don't need to for the logic )